### PR TITLE
Bdog 1283

### PR DESCRIPTION
**Pull request by JasperWagen**

### CODE REVIEW

File: app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html [hunk 1]
```diff
@@ -136,3 +143,9 @@ <h1>What's Running Where? @platform.displayName</h1>
     </li>
   }
 }
+
+@matchesStaging(wrw: WhatsRunningWhere, version: WhatsRunningWhereVersion) = @{
+    version.versionNumber.asString==wrw.versions
```
Comment (colin-lamed) [inline app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html:148]: can you not compare VersionNumbers without converting to String?

**Comment by JasperWagen** [comment]
(https://user-images.githubusercontent.com/68235822/106748562-5325c100-661d-11eb-95ee-795e265da38d.png)

File: conf/app.routes [hunk 2]
```diff
@@ -62,7 +62,7 @@ GET        /shuttering/:shutterType/:env                uk.gov.hmrc.cataloguefro
 GET        /shutter-events                              uk.gov.hmrc.cataloguefrontend.shuttering.ShutterEventsController.shutterEvents
 GET        /shutter-events/list                         uk.gov.hmrc.cataloguefrontend.shuttering.ShutterEventsController.shutterEventsList(env: Environment, serviceName: Option[String] ?= None)
 
-GET        /whats-running-where-heritage                uk.gov.hmrc.cataloguefrontend.whatsrunningwhere.WhatsRunningWhereController.heritageReleases
-GET        /whats-running-where                        uk.gov.hmrc.cataloguefrontend.whatsrunningwhere.WhatsRunningWhereController.ecsReleases
+GET        /whats-running-where-heritage                uk.gov.hmrc.cataloguefrontend.whatsrunningwhere.WhatsRunningWhereController.heritageReleases(showDiff: Boolean ?= false)
```
Comment (anshulbajpai) [inline conf/app.routes:65]: How will this feature be discovered if the query param has a default value? The WRW page can have a button to toggle this flag.

**Comment by anshulbajpai** [comment]
> (https://user-images.githubusercontent.com/68235822/106748562-5325c100-661d-11eb-95ee-795e265da38d.png)

It's not clear what that yellow colour means. We can add some kind of legend or help text to help users understand what colours mean.

File: public/catalogue-frontend.css [hunk 3]
```diff
@@ -12376,4 +12376,17 @@ table.shutter-events {
   transition: height .5s;
   overflow: hidden;
   height:120px; /* If you change this, change the corresponding smallHeight value in catalogue-frontend.css too */
+}
+
+.env-maj-version-behind{
```
Comment (cjwebb) [inline public/catalogue-frontend.css:12381]: Is it any version behind, not just major versions?

File: app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html [hunk 4]
```diff
@@ -136,3 +143,9 @@ <h1>What's Running Where? @platform.displayName</h1>
...
+            .find(_.environment==Environment.Staging)
```
Comment (anshulbajpai) [inline app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html:149]: Are we comparing `String` and `Any` here?

File: app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html [hunk 5]
```diff
@@ -89,7 +90,13 @@ <h1>What's Running Where? @platform.displayName</h1>
 
                 @environments.map { env =>
                     @wrw.versions.find(_.environment == env).map { version =>
-                        <td id="row@{i}_version_@{env.asString}">@version.versionNumber.asString</td>
+                        <td id="row@{i}_version_@{env.asString}"
+                        @if(showDiff && env == Environment.Production) {
+                            @if(!matchesStaging(wrw, version)){
```
Comment (anshulbajpai) [inline app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html:95]: I don't think we should add this class when prod is ahead of staging or if staging is missing. At the moment `matchesStaging ` function is doing equality check.

**Comment by anshulbajpai** [comment]
How can we test this?

File: conf/app.routes (hunk 2 above)
Comment (JasperWagen) [reply conf/app.routes:65]: Feature, not a bug! - it's in order to stealth release to get feedback before making discoverable

File: public/catalogue-frontend.css (hunk 3 above)
Comment (JasperWagen) [reply public/catalogue-frontend.css:12381]: Yeah it's any version, good spot

File: app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html (hunk 5 above)
Comment (anshulbajpai) [reply app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html:95]: Or is that the intention to show a visual difference if there is any version difference.

File: app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html (hunk 1 above)
Comment (JasperWagen) [reply app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html:148]: This is a good point, I will try without.

File: app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html (hunk 5 above)
Comment (JasperWagen) [reply app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html:95]: Yes so the acceptance criteria for this were to highlight if(stagingVersion!=productionVersion), I will however get feedback on making that if(stagingVersion>productionVersion)

The case of staging being missing is something I haven't considered however, will look into that.

File: app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html (hunk 1 above)
Comment (anshulbajpai) [inline app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html:148]: Are we comparing `String` and `Any` here?

File: app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html [hunk 6]
```diff
@@ -136,3 +143,11 @@ <h1>What's Running Where? @platform.displayName</h1>
     </li>
   }
 }
+
+@matchesStaging(wrw: WhatsRunningWhere, prodVersion: WhatsRunningWhereVersion) = @{
+    wrw.versions.find(_.environment==Environment.Staging).map(_.versionNumber) match {
+        case Some(prodVersion.versionNumber) => true
```
[the remark this answers is not in the document]
Comment (JasperWagen) [reply app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html:149]: Okay, thanks. Could I put a getOrElse in the inline statement? ie:

```  
    wrw.versions.find(_.environment == Environment.Staging).map(_.versionNumber).getOrElse(None) match {
        case prodVersion.versionNumber => true
        case None => true
        case _ => false
    }
```

Or is this ill-advised?

[the remark this answers is not in the document]
Comment (anshulbajpai) [reply app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html:149]: I guess you can do something like 
```scala
wrw.versions.find(_.environment == Environment.Staging)
.map(_.versionNumber)
.map(_ == prodVersion.versionNumber)
.getOrElse(true)
```

Although,  the function `matchesStaging ` is quite weird. It returns `true` even when staging version is absent. How can it be **matching** in that case? 😄  

If it returns false to just make it consistent with the semantics of being different then probably you can do something like 
```scala
wrw.versions.find(_.environment == Environment.Staging)
.map(_.versionNumber)
.exists(_ == prodVersion.versionNumber)
```

[the remark this answers is not in the document]
Comment (JasperWagen) [reply app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html:149]: I guess the point of the none case returning true is that it assumes that having no staging version is fine and does not require highlighting, I can also see the argument for the opposite assumption. Think this is one to ask the users about. 

The other reason I moved to a match statement is to expand in the future, say if more than one environment needed to be matched against.

[the remark this answers is not in the document]
Comment (anshulbajpai) [reply app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html:149]: We can refactor this code later to accommodate future changes (if needed). It is not going to be complicated refactoring. Let's keep it simple for now and use more functional combinators.

[the remark this answers is not in the document]
Comment (JasperWagen) [reply app/views/whatsrunningwhere/WhatsRunningWherePage.scala.html]: Users have fedback that highlighting in the above case would be useful so I have refactored and pushed

File: app/uk/gov/hmrc/cataloguefrontend/whatsrunningwhere/WhatsRunningWhereController.scala [hunk 7]
```diff
@@ -51,21 +53,30 @@ class WhatsRunningWhereController @Inject()(
...
+          ).mapN { (r, p) => (r, p) }
+        environments = distinctEnvironments(releases)
+        profileNames = profiles.filter(_.profileType == selectedProfileType).map(_.profileName).sorted
+      } yield Ok(page(platform, environments, releases, selectedProfileType, profileNames, form, showDiff))
     }
 }
 
+object WhatsRunningWhereController {
+  def matchesProduction(wrw: WhatsRunningWhere, prodVersion: WhatsRunningWhereVersion, comparedEnv: Environment): Boolean = {
+    println(wrw.applicationName, prodVersion)
```
Comment (anshulbajpai) [inline app/uk/gov/hmrc/cataloguefrontend/whatsrunningwhere/WhatsRunningWhereController.scala:73]: println can be removed

File: test/uk/gov/hmrc/cataloguefrontend/whatsrunningwhere/WhatsRunningWhereControllerSpec.scala [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    val wrwVersionStagingNotMatchesProd = WhatsRunningWhereVersion(Environment.Staging, Platform.ECS, versionNumberNotProd, timeSeen)
+
+    val wrwMatchesProd = WhatsRunningWhere(ServiceName("foo"), List(wrwVersionProd, wrwVersionStagingMatchesProd))
+    val wrwNotMatchesProd = WhatsRunningWhere(ServiceName("foo"), List(wrwVersionProd, wrwVersionStagingNotMatchesProd))
+
+    "return true when compared env version matches production version" in {
+      WhatsRunningWhereController.matchesProduction(wrwMatchesProd, wrwVersionProd, Environment.Staging) shouldBe true
+    }
+
+    "return false when compared env version matches production version" in {
```
Comment (anshulbajpai) [inline test/uk/gov/hmrc/cataloguefrontend/whatsrunningwhere/WhatsRunningWhereControllerSpec.scala:27]: The description doesn't match the test intent.

**Review by anshulbajpai — APPROVED** [review]

**Review by anshulbajpai — APPROVED** [review]

**Pull request merged**